test/e2e/clusterctl_upgrade.go

```diff
@@ -263,6 +277,22 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			})
 
 			if !input.SkipCleanup {
+				if discovery.ServerSupportsVersion(managementClusterProxy.GetClientSet().DiscoveryClient, clusterv1.GroupVersion) == nil {
+					Byf("Deleting all clusters in Namespace: %s in management cluster: %s", testNamespace.Name, managementClusterName)
+					framework.DeleteAllClustersAndWait(ctx, framework.DeleteAllClustersAndWaitInput{
+						Client:    managementClusterProxy.GetClient(),
+						Namespace: testNamespace.Name,
+					}, input.E2EConfig.GetIntervals(specName, "wait-delete-cluster")...)
+				} else if discovery.ServerSupportsVersion(managementClusterProxy.GetClientSet().DiscoveryClient, clusterv1old.GroupVersion) == nil {
+					Byf("Deleting all clusters in Namespace: %s in management cluster: %s", testNamespace.Name, managementClusterName)
+					deleteAllClustersAndWait(ctx, framework.DeleteAllClustersAndWaitInput{
+						Client:    managementClusterProxy.GetClient(),
+						Namespace: testNamespace.Name,
+					}, input.E2EConfig.GetIntervals(specName, "wait-delete-cluster")...)
+				} else {
+					log.Logf("Management Cluster does not appear to support CAPI resources.")
+				}
```

Comment:
This cleanup step would previously fail when the spec fails prior to upgrading the management cluster, this allows cleanup to proceed if failure occurs prior to running `clusterctl upgrade`, or even before `clusterctl init` has been run.